### PR TITLE
Implement new bg estimation

### PR DESCRIPTION
**Pull request by constantinpape**

### CODE REVIEW

**Comment by constantinpape** [comment]
The overall structure is there.
What's left to do:
- Add `background_fraction` to result of `ExtractBackground`
- test and debug ...

**Comment by constantinpape** [comment]
> Side note: If we encounter performance issues there, we can optimize the voronoi segmentation by a lot I think.

I think it's super fast already; < 5 min for the whole plate, so not an issue for now.

**Comment by constantinpape** [comment]
> > Side note: If we encounter performance issues there, we can optimize the voronoi segmentation by a lot I think.
> 
> I think it's super fast already; < 5 min for the whole plate, so not an issue for now.

@imagirom ok, I was wrong. For large dilations it actually takes quite long.
What did you have in mind to speed it up?

**Comment by constantinpape** [comment]
Ok, I made the new bg estimation work, but still need to properly test it and then run it for all the plates.
@imagirom if you have time, could you have another look at my changes?

Beside from testing, what's still left to do:
- Make the min background estimation robust against fully empty plates
- Speed up voronoi (happy to hear your ideas, @imagirom)
- Implement segmentation properties, see #128
- Add nucleus size based cell level qc params 

Also, the intensity based qc is not implemented yet (I will probably still add the voronoi + feature extraction here, but add the QC in a separate PR). For this, we need to
- Add another voronoi task (with small dilation and excluding nucleus) + extract features from it
- Load the result in in `ImageLevelQC`, implement criterion to find intensity outliers
- Make the excel table export for dilated nuclei work again so that @tischi can plot histograms so we can determine a good criterion

**Comment by imagirom** [comment]
> For large dilations it actually takes quite long.
> What did you have in mind to speed it up?

I think that the line `ring_mask = morph.dilation(input_mask, morph.disk(dilation_radius))` is the only part whose runtime depends on the dilation radius. And it should be possible to replace it by a threshold on the distance transform that is computed anyways for the watershed.
(https://github.com/hci-unihd/batchlib/blob/new-bg-estimation/batchlib/segmentation/voronoi_ring_segmentation.py#L72)

Also, if the voronoi segmentation is computed for multiple radii, we could compute the distance transform only once.

**Comment by imagirom** [comment]
> @imagirom if you have time, could you have another look at my changes?

I do not have time right now, but will do later

**Comment by constantinpape** [comment]
> I think that the line `ring_mask = morph.dilation(input_mask, morph.disk(dilation_radius))` is the only part whose runtime depends on the dilation radius. And it should be possible to replace it by a threshold on the distance transform that is computed anyways for the watershed.
> (https://github.com/hci-unihd/batchlib/blob/new-bg-estimation/batchlib/segmentation/voronoi_ring_segmentation.py#L72)

Ok, that makes a lot of sense.

> I do not have time right now, but will do later

I also don't have time to work on this any more today, I will have another look tomorrow and looking forward to hear your thoughts.

File: batchlib/analysis/background_extraction.py [hunk 1]
```diff
@@ -120,53 +122,76 @@ def get_bg_stats(self, bg_values):
...
                 logger.info(f'{self.name}: skipping outlier image in bg calculation')
                 continue
+
             bg_per_well_dict[well].append(bg_segment)
+            bg_fractions_per_well[well].append(bg_fractions_per_image[image_name])
+
         bg_per_well_dict = {well: torch.cat(bg_segments, dim=1) for well, bg_segments in bg_per_well_dict.items()}
         bg_per_well_stats = {well: self.get_bg_stats(bg_per_well_dict.get(well, None)) for well in wells}
+        bg_fractions_per_well = {well: np.median(bg_fractions_per_well[well]) if well in bg_fractions_per_well
+                                 else None for well in wells}
```
Comment (imagirom) [inline batchlib/analysis/background_extraction.py:184]: If think this should be the mean (at least as long as all images have the same size, that should be the right way to compute the bg fraction over all images per well).

File: batchlib/analysis/background_extraction.py [hunk 2]
```diff
@@ -120,53 +122,76 @@ def get_bg_stats(self, bg_values):
...
 
         bg_plate_stats = self.get_bg_stats(torch.cat(list(bg_per_well_dict.values()), dim=1)
                                            if len(bg_per_well_dict) > 0 else None)
+        plate_bg_fraction = [frac for frac in bg_fractions_per_well.values() if frac is not None]
+        plate_bg_fraction = np.median(plate_bg_fraction) if len(plate_bg_fraction) > 0 else None
```
Comment (imagirom) [inline batchlib/analysis/background_extraction.py:189]: I think taking the median is totally fine to get an idea of the amount of BG in the plate.
But to really get the fraction of BG pixels in non-outlier images in the entire plate, somethig like 
`plate_bg_fraction = np.mean(np.concatenate([bg_fractions for bg_fractions in bg_fractions_per_well.values() if bg_fractions is not None]))`
should be correct.

File: batchlib/workflows/cell_analysis.py [hunk 3]
```diff
@@ -335,10 +322,8 @@ def core_workflow_tasks(config, name, feature_identifier):
 
     # we could also just add these on demand depending on what we need according to the background params
     bg_estimation_keys = [marker_ana_in_key] + serum_ana_in_keys
-    job_list, bg_tables1 = add_background_estimation_from_min_well(job_list, config,
-                                                                   bg_estimation_keys, feature_identifier)
-    job_list, bg_tables2 = add_background_estimation(job_list, config, config.seg_key,
-                                                     bg_estimation_keys, feature_identifier)
+    job_list = add_background_estimation_from_min_well(job_list, config, bg_estimation_keys)
+    job_list = add_background_estimation(job_list, config.seg_key, bg_estimation_keys)
```
Comment (imagirom) [inline batchlib/workflows/cell_analysis.py:326]: This is just personal preference, but all the `add_*` methods act on the `job_list` in place, right? Then I would maybe remove the `job_list = ` to make this more obvious.

File: batchlib/analysis/background_extraction.py (hunk 1 above)
Comment (constantinpape) [reply batchlib/analysis/background_extraction.py:184]: Yes, makes sense.

File: batchlib/analysis/background_extraction.py (hunk 2 above)
Comment (constantinpape) [reply batchlib/analysis/background_extraction.py:189]: Yes, I agree. mean is more appropriate.

File: batchlib/workflows/cell_analysis.py (hunk 3 above)
Comment (constantinpape) [reply batchlib/workflows/cell_analysis.py:326]: Yes, they do work in place; but I usually prefer to give it as explicit return value too.
(The implicit reference / not reference arguments is one of my pet peeves with python, so I mostly try to be explicit with return value in cases where it doesn't make the code more unreadable.)

**Comment by constantinpape** [comment]
> Apart from comments below, looks good to me. 

Thanks!

> Especially the changes to the Analysis base class^^

Yes, much better this way :)

**Comment by tischi** [comment]
> And it should be possible to replace it by a threshold on the distance transform that is computed anyways for the watershed.

I am just curious: Is this also faster when the distance transform would not have been precomputed? If so, why is this not the default implementation for dilations? I am asking because, e.g., morphological closings in imglib2 are sometime suuuper slow.

**Comment by constantinpape** [comment]
@imagirom I addressed your comments and implemented the dilation based on thresholding the distance transform, seems to work and def. much faster for large dilation radii.
Will still clean up a few minor things and then rerun.

@tischi:
> I am just curious: Is this also faster when the distance transform would not have been precomputed? If so, why is this not the default implementation for dilations? I am asking because, e.g., morphological closings in imglib2 are sometime suuuper slow.

Yes, I think there is some dilation radius when doing it via the distance transform becomes faster.
Not sure when exactly that is though.

**Pull request merged**